### ISSUE TITLE
Tensorboard callback not logging learning rate when not using LearningRateSchedule (e.g. when using ReduceLROnPlateau)

### ISSUE THREAD

**Issue by MatejHl** [tensorflow/tensorboard#4522]
### Diagnostics

<details>
<summary>Diagnostics output</summary>

``````
--- check: autoidentify
INFO: diagnose_tensorboard.py version 5e810a97d96ec99b9b6b596476f2883baa7c22b6

--- check: general
INFO: sys.version_info: sys.version_info(major=3, minor=8, micro=5, releaselevel='final', serial=0)
INFO: os.name: posix
INFO: os.uname(): posix.uname_result(sysname='Linux', nodename='matej-Precision-7920-Tower', release='4.15.0-1027-oem', version='#32-Ubuntu SMP Sat Nov 10 06:29:08 UTC 2018', machine='x86_64')
INFO: sys.getwindowsversion(): N/A

--- check: package_management
INFO: has conda-meta: True
INFO: $VIRTUAL_ENV: None

--- check: installed_packages
INFO: installed: tensorboard==2.3.0
INFO: installed: tf-nightly==2.5.0.dev20210104
INFO: installed: tensorflow-estimator==2.4.0
INFO: installed: tf-estimator-nightly==2.5.0.dev2021010401
WARNING: conflicting installations: ['tensorflow-estimator', 'tf-estimator-nightly']

--- check: tensorboard_python_version
INFO: tensorboard.version.VERSION: '2.3.0'

--- check: tensorflow_python_version
2021-01-07 14:33:18.900710: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
INFO: tensorflow.__version__: '2.5.0-dev20210104'
INFO: tensorflow.__git_version__: 'v1.12.1-48302-gcbf45b6908f'

--- check: tensorboard_binary_path
INFO: which tensorboard: b'/home/matej/miniconda3/envs/chemModel/bin/tensorboard\n'

--- check: addrinfos
socket.has_ipv6 = True
socket.AF_UNSPEC = <AddressFamily.AF_UNSPEC: 0>
socket.SOCK_STREAM = <SocketKind.SOCK_STREAM: 1>
socket.AI_ADDRCONFIG = <AddressInfo.AI_ADDRCONFIG: 32>
socket.AI_PASSIVE = <AddressInfo.AI_PASSIVE: 1>
Loopback flags: <AddressInfo.AI_ADDRCONFIG: 32>
Loopback infos: [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::1', 0, 0, 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 0))]
Wildcard flags: <AddressInfo.AI_PASSIVE: 1>
Wildcard infos: [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('0.0.0.0', 0)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::', 0, 0, 0))]

--- check: readable_fqdn
INFO: socket.getfqdn(): 'matej-Precision-7920-Tower'

--- check: stat_tensorboardinfo
INFO: directory: /tmp/.tensorboard-info
INFO: os.stat(...): os.stat_result(st_mode=16895, st_ino=243008571, st_dev=2051, st_nlink=2, st_uid=1001, st_gid=1001, st_size=4096, st_atime=1609924616, st_mtime=1610026321, st_ctime=1610026321)
INFO: mode: 0o40777

--- check: source_trees_without_genfiles
INFO: tensorboard_roots (1): ['/home/matej/miniconda3/envs/chemModel/lib/python3.8/site-packages']; bad_roots (0): []

--- check: full_pip_freeze
INFO: pip freeze --all:
absl-py==0.11.0
astunparse==1.6.3
cachetools==4.2.0
certifi==2020.12.5
chardet==4.0.0
cycler==0.10.0
decorator==4.4.2
et-xmlfile==1.0.1
flatbuffers==1.12
gast==0.3.3
google-auth==1.24.0
google-auth-oauthlib==0.4.2
google-pasta==0.2.0
grpcio==1.34.0
gviz-api==1.9.0
h5py==3.1.0
idna==2.10
jdcal==1.4.1
joblib==1.0.0
Keras-Preprocessing==1.1.2
kiwisolver @ file:///home/conda/feedstock_root/build_artifacts/kiwisolver_1604322295622/work
lxml==4.6.2
Markdown==3.3.3
matplotlib @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-suite_1605180228501/work
-e git+https://github.com/MatejHl/Mol2Graph.git@89206d3405123de5c56b300b9e72fa1d3eb671f8#egg=mol2graph
networkx==2.5
numpy @ file:///home/conda/feedstock_root/build_artifacts/numpy_1607937295372/work
oauthlib==3.1.0
olefile @ file:///home/conda/feedstock_root/build_artifacts/olefile_1602866521163/work
openpyxl==3.0.5
opt-einsum==3.3.0
pandas==1.2.0
Pillow @ file:///home/conda/feedstock_root/build_artifacts/pillow_1609627771882/work
pip==20.3.3
protobuf==3.14.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycairo==1.20.0
pyparsing==2.4.7
python-dateutil==2.8.1
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1608904108784/work
reportlab==3.5.57
requests==2.25.1
requests-oauthlib==1.3.0
rsa==4.6
scikit-learn==0.24.0
scipy==1.6.0
setuptools==51.0.0.post20201207
six @ file:///home/conda/feedstock_root/build_artifacts/six_1590081179328/work
spektral==1.0.3
SQLAlchemy @ file:///home/conda/feedstock_root/build_artifacts/sqlalchemy_1608366814204/work
tensorboard==2.3.0
tensorboard-plugin-profile==2.4.0
tensorboard-plugin-wit==1.7.0
tensorflow-addons==0.12.0
tensorflow-estimator==2.4.0
termcolor==1.1.0
tf-estimator-nightly==2.5.0.dev2021010401
tf-nightly==2.5.0.dev20210104
threadpoolctl==2.1.0
tornado @ file:///home/conda/feedstock_root/build_artifacts/tornado_1604105045397/work
tqdm==4.55.1
typeguard==2.10.0
typing-extensions==3.7.4.3
urllib3==1.26.2
Werkzeug==1.0.1
wheel==0.36.2
wrapt==1.12.1

``````

</details>

## Issue description
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Linux Ubuntu 18.04
- TensorFlow installed from: pip3
- TensorFlow version: tf-nightly 2.5.0-dev20210104
- Python version: 3.8.5
- CUDA/cuDNN version: 11.0.3 / 8.0.4
- GPU model and memory: Quadro RTX 5000 / 16384 MB

**Describe the current behavior**  
If there is no LearningRateSchedule (either LR is constant or is changed via callback) then [Tensorboard callback](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/TensorBoard) doesn't log learning rate. This is especially relevant when using [ReduceLROnPlateau](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/ReduceLROnPlateau) callback.

**Describe the expected behavior**  
Learning rate should be always logged or at least in the case of using a callback.

**Standalone code to reproduce the issue**  
```
import tensorflow as tf
import datetime
import os

class SimpleModel(tf.keras.Model):
    def __init__(self):
        super(SimpleModel, self).__init__()
        self.Dense_in = tf.keras.layers.Dense(units = 50, 
                                activation=tf.nn.relu, 
                                use_bias=True)
        self.Dense_hidden = tf.keras.layers.Dense(units = 20, 
                                activation=tf.nn.relu, 
                                use_bias=True)
        self.Dense_out = tf.keras.layers.Dense(units = 1, 
                                activation=None, 
                                use_bias=True)

    def call(self, inputs, training = True):
        X = self.Dense_in(inputs)
        X = self.Dense_hidden(X)
        X = self.Dense_out(X)
        return X

dummy_data = tf.random.normal((10000, 50))
dummy_y = tf.random.uniform((10000,), minval=0, maxval=2, dtype=tf.dtypes.int32)


#### LR is not logged here ####
model = SimpleModel()

loss = tf.keras.losses.BinaryCrossentropy(name='binary_crossentropy', from_logits=True)
opt = tf.keras.optimizers.Adam(learning_rate = 0.001)
model.compile(optimizer = opt, loss = loss, metrics=["acc"])

_datetime = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
log_dir = os.path.join('dummy_logs', _datetime)
tb_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir)

model.fit(x = dummy_data, y = dummy_y, epochs = 10,
        callbacks=[tf.keras.callbacks.ReduceLROnPlateau(monitor='acc', 
                        factor=0.5, patience=10, verbose=0, mode='auto', 
                        min_delta=0.0001, cooldown=0, min_lr=0),
                    tb_callback])

#### LR is logged here ####
model = SimpleModel()

lr_schedule = tf.keras.optimizers.schedules.ExponentialDecay(
    initial_learning_rate = 0.001,
    decay_steps=100000,
    decay_rate=0.96)

loss = tf.keras.losses.BinaryCrossentropy(name='binary_crossentropy', from_logits=True)
opt = tf.keras.optimizers.Adam(learning_rate = lr_schedule)
model.compile(optimizer = opt, loss = loss, metrics=["acc"])

_datetime = datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
log_dir = os.path.join('dummy_logs', _datetime)
tb_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir)

model.fit(x = dummy_data, y = dummy_y, epochs = 10,
        callbacks=[tb_callback])
```
![Screenshot from 2021-01-07 12-14-15](https://user-images.githubusercontent.com/32313554/103887434-517ee100-50e3-11eb-9597-4978bee27972.png)

**Other info / logs**   
I think that the problem is in tf.keras.callbacks.TensorBoard in _collect_learning_rate (see code block below). If self.model.optimizer.lr is sheduler then logs['learning_rate'] is created and LR can be seen in tensorboard. However if self.model.optimizer.lr is not an instance of learning_rate_schedule.LearningRateSchedule then logs['learning_rate'] does not exists and thus LR is not logged.
```
def _collect_learning_rate(self, logs):
    lr_schedule = getattr(self.model.optimizer, 'lr', None)
    if isinstance(lr_schedule, learning_rate_schedule.LearningRateSchedule):
      logs['learning_rate'] = lr_schedule(self.model.optimizer.iterations)
    return logs
```

A quick fix that I see here could be to add *else* or *elif* and directly log self.model.optimizer.lr. Changing as below fixed the issue for me when using ReduceLROnPlateau callback. This fix assumes that if no LearningRateSchedule is used then self.model.optimizer.lr could be logged as scalar by tensorboard callback. If this is not always possible then I guess adding some reasonable *elif* to be able to catch LR that can be logged (e.g. scalar) could resolve the problem.
```
class TensorBoardCallback(tf.keras.callbacks.TensorBoard):
    def __init__(self, **kwargs):
        super(TensorBoardCallback, self).__init__(**kwargs)

    def _collect_learning_rate(self, logs):
        lr_schedule = getattr(self.model.optimizer, 'lr', None)
        if isinstance(lr_schedule, tf.keras.optimizers.schedules.LearningRateSchedule):
            logs['learning_rate'] = lr_schedule(self.model.optimizer.iterations)
        elif lr_schedule is not None:
            logs['learning_rate'] = lr_schedule
        return logs
```

**Comment by rchao** [comment]
Hello @MatejHl, thanks for reporting the issue. Would you be interested in contributing a PR to fix?

**Comment by MatejHl** [comment]
Hello @rchao sorry for a bit late reply. 

Yes I'd be very happy to contribute.

The solution above assumes that nothing "ugly" that can't be logged via tf.summary.scalar can be passed as self.model.optimizer.lr (i.e. nothing not convertible to Tensor). I guess there would be an error raised somewhere in Optimizer itself during backpropagation if this "ugly" learning rate is used, but could you please confirm this guess?

**Comment by rchao** [comment]
No problem, thanks for replying @MatejHl, and we'd appreciate your contribution. To answer your question, I would simply give it a try by having an "ugly" value for self.model.optimizer.lr and attempting to log it. There are guesses but actual testing would be the best to confirm :)

**Comment by MatejHl** [comment]
PR https://github.com/tensorflow/tensorflow/pull/46571 merged some time ago, so I am closing this issue. 

Have a nice day and be happy.

**Comment by matangover** [comment]
The PR that closed this issue (https://github.com/tensorflow/tensorflow/pull/46571) was reverted on the same day that it was merged. What was the reason? Can this issue be re-opened?
